python/lsst/daf/butler/core/dimensions/coordinate.py

```
@@ -0,0 +1,260 @@
+        try:
+            values = tuple(d[name] for name in graph.required.names)
+        except KeyError as err:
+            raise KeyError(f"No value in data ID for required dimension {err}.") from err
+        return DataCoordinate(graph, values)
+
+    def byName(self) -> Dict[str, Any]:
+        """Return a true `dict` keyed by `str` dimension name and the same
+        values as ``self``.
+        """
+        return {k.name: v for k, v in self.items()}
+
+    def __getnewargs__(self) -> tuple:
+        # Implements pickle support (in addition to methods provided by
+        # @immutable decorator).
+        return (self.graph, self.values())
+
+    def __hash__(self) -> int:
+        return hash((self.graph, self.values()))
+
+    def __eq__(self, other: DataCoordinate) -> bool:
+        try:
+            # Optimized code path for DataCoordinate comparisons.
+            return self.graph == other.graph and self.values() == other.values()
+        except AttributeError:
+            # Also support comparison with informal data ID dictionaries that
+            # map dimension name to value.
+            return self.byName() == other
+
+    def __str__(self):
+        return f"{dict(self.byName())}"
+
+    def __repr__(self):
+        return f"DataCoordinate({self.graph}, {self.values()})"
+
+    def fingerprint(self, update):
+        """Update a secure hash function with the values in this data ID.
+
+        Parameters
+        ----------
+        update : `~collections.abc.Callable`
+            Callable that accepts a single `bytes` argument to update
+            the hash; usually the ``update`` method of an instance from
+            the ``hashlib`` module.
+        """
+        for k, v in self.items():
+            update(k.name.encode("utf8"))
+            if isinstance(v, int):
+                update(v.to_bytes(64, "big", signed=False))
+            elif isinstance(v, str):
+                update(v.encode("utf8"))
```

No other types of dimensions could ever exist besides int and str? Maybe throw an exception for an unexpected type here?